helpers/kernel_symbols.go

```diff
@@ -3,340 +3,315 @@ package helpers
+	go k.updateSymbolMap(symbolChan)
+	go k.updateAddrsMap(addrChan)
+	go k.updateSymByNameMap(symByNameChan)
+	go k.updateSymByAddrMap(symByAddrChan)
+
+	// Send kallsyms lines to the map update goroutines.
+	err := k.processLines([]chan *KernelSymbol{
+		symbolChan,
+		addrChan,
+		symByNameChan,
+		symByAddrChan,
+	})
 	if err != nil {
 		return err
 	}
-	k.textSegStart = etext.Address
-	return nil
-}
 
-// lazyKernelSymbols
-type lazyKernelSymbols struct {
-	fileContent   []string
-	symbolMap     map[string]*KernelSymbol
-	symbolAddrMap map[uint64]*KernelSymbol
-	textSegStart  uint64
-	textSegEnd    uint64
-}
+	// Finally, wait for the map update goroutines to finish.
+	k.updateWg.Wait()
 
-// NewLazyKernelSymbolsMap will return a lazy implementation of the KernelSymbolTable
-// The lazy implementation keeps a copy of the /proc/kallsyms file content and queries that
-// copy on demand, instead of preparsing it.
-// It keeps caches of previously found results.
-func NewLazyKernelSymbolsMap() (KernelSymbolTable, error) {
-	k := &lazyKernelSymbols{}
-	err := k.Refresh()
-	return k, err
+	// Get the kernel text segment addresses.
+	return k.getTextSegmentAddresses()
 }
 
-func (k *lazyKernelSymbols) TextSegmentContains(addr uint64) (bool, error) {
-	// query the segments if not queried yet
-	if k.textSegEnd < k.textSegStart {
-		stext, err := k.GetSymbolByName("system", "_stext")
-		if err != nil {
-			return false, err
-		}
-		k.textSegStart = stext.Address
-		etext, err := k.GetSymbolByName("system", "_etext")
-		if err != nil {
-			return false, err
-		}
-		k.textSegEnd = etext.Address
+//
+// Private methods.
+//
+
+// getTextSegmentAddresses gets the start and end addresses of the kernel text segment.
+func (k *KernelSymbolTable) getTextSegmentAddresses() error {
+	stext, exist1 := k.symByName[nameAndOwner{"_stext", "system"}]
+	etext, exist2 := k.symByName[nameAndOwner{"_etext", "system"}]
+
+	if !exist1 || !exist2 {
+		return fmt.Errorf("kernel text segment symbol(s) not found")
 	}
-	return addr >= k.textSegStart && addr < k.textSegEnd, nil
+
+	k.textSegStart = stext[0].Address
+	k.textSegEnd = etext[0].Address
+
+	return nil
 }
 
-func (k *lazyKernelSymbols) GetSymbolByName(owner string, name string) (*KernelSymbol, error) {
-	key := symbolKey(owner, name)
-	symbol, exist := k.symbolMap[key]
-	if exist {
-		return symbol, nil
+//
+// Concurrency logic for updating the maps
+//
+
+// processLines process lines from kallsyms and sends them to map update goroutines.
+func (k *KernelSymbolTable) processLines(chans []chan *KernelSymbol) error {
+	file, err := os.Open(kallsymsPath)
+	if err != nil {
+		return err
 	}
-	for _, line := range k.fileContent {
-		line := strings.Fields(line)
-		// if the line is less than 3 words, we can't parse it (one or more fields missing)
-		// if the searched owner isn't system and the line counter is less than 4 words, the line is irrelevant
-		if len(line) < 3 || (owner != "system" && len(line) < 4) {
```

Comment:
Not sure, but I dont think all kallsyms variations were considered when this package was created and this was a generic attempt of not failing. I think the original idea was possibly to avoid the final lines, coming from modules (thus not "system"), to break something, but can't tell for sure and haven't seen a real scenario it was needed.